crates/biome_resolver/src/lib.rs

```diff
@@ -0,0 +1,589 @@
+            }
+            None => resolve_string(target.as_str()),
+        },
+        _ => Err(ResolveError::InvalidMappingTarget),
+    }
+}
+
+fn resolve_dependency(
+    specifier: &str,
+    base_dir: &Utf8Path,
+    fs: &dyn ResolverFsProxy,
+    options: &ResolveOptions,
+) -> Result<Utf8PathBuf, ResolveError> {
+    let (package_name, subpath) = parse_package_specifier(specifier)?;
+
+    for dir in base_dir.ancestors() {
+        let package_path = Utf8PathBuf::from(format!("{dir}/node_modules/{package_name}"));
+        let package_path = match fs.path_info(&package_path) {
+            Ok(PathInfo::Directory) => package_path,
+            Ok(PathInfo::Symlink { normalized_target }) => normalized_target,
+            _ => continue,
+        };
+
+        if let Ok(package_json) = fs.read_package_json(&package_path.join("package.json")) {
+            if package_json.get_value_by_path(&["exports"]).is_some() {
+                return resolve_export(subpath, &package_path, &package_json, fs, options);
+            }
+
+            if subpath.is_empty() {
+                if let Some(main_target) = package_json
+                    .get_value_by_path(&["main"])
+                    .and_then(JsonValue::as_string)
+                {
+                    let options = options.without_extensions_or_manifests();
+                    return resolve_relative_path(
+                        main_target.as_str(),
+                        &package_path,
+                        fs,
+                        &options,
+                    );
+                }
+            }
+        }
+
+        let options = options.without_extensions_or_manifests();
+        return resolve_relative_path(subpath, &package_path, fs, &options);
+    }
+
+    Err(ResolveError::NotFound)
+}
+
+enum ResolvedPathInfo {
+    Directory,
+    File,
+}
+
+/// Resolves the given `path` to a tuple of [`ResolvedPathInfo`] and the real
+/// path being pointed to.
+fn resolve_path_info(
+    path: Utf8PathBuf,
+    fs: &dyn ResolverFsProxy,
+) -> Result<(ResolvedPathInfo, Utf8PathBuf), ResolveError> {
+    match fs.path_info(&path)? {
+        PathInfo::Directory => Ok((ResolvedPathInfo::Directory, path)),
+        PathInfo::File => Ok((ResolvedPathInfo::File, path)),
+        PathInfo::Symlink { normalized_target } => match fs.path_info(&normalized_target)? {
+            PathInfo::Directory => Ok((ResolvedPathInfo::Directory, normalized_target)),
+            PathInfo::File => Ok((ResolvedPathInfo::File, normalized_target)),
+            PathInfo::Symlink { .. } => Err(ResolveError::BrokenSymlink),
+        },
+    }
+}
+
+fn normalize_owned_path(path: Utf8PathBuf) -> Utf8PathBuf {
+    let string = path.as_str();
+    if string.len() < 3 {
+        return path;
+    }
+
+    let mut prev = string.chars().next().unwrap();
+    for byte in string.chars().skip(1) {
+        if byte == '.' && (prev == '.' || prev == std::path::MAIN_SEPARATOR) {
```

Comment:
Do you think that using bytes instead of characters could improve performance?

Reply:
I actually tried, but it became quite messy because `std::path::MAIN_SEPARATOR` is a `char`, so I left it at this.